neo/Network/P2P/Payloads/VersionPayload.cs

```diff
@@ -1,67 +1,72 @@
 ﻿using Neo.IO;
+using Neo.Network.P2P.Capabilities;
 using System;
+using System.Collections.Generic;
 using System.IO;
+using System.Linq;
 
 namespace Neo.Network.P2P.Payloads
 {
     public class VersionPayload : ISerializable
     {
+        public const int MaxCapabilities = 32;
+
         public uint Magic;
         public uint Version;
-        public VersionServices Services;
         public uint Timestamp;
-        public ushort Port;
+        public NodeCapabilityBase[] Capabilities;
         public uint Nonce;
         public string UserAgent;
-        public uint StartHeight;
 
         public int Size =>
-            sizeof(uint) +              //Magic
-            sizeof(uint) +              //Version
-            sizeof(VersionServices) +   //Services
-            sizeof(uint) +              //Timestamp
-            sizeof(ushort) +            //Port
-            sizeof(uint) +              //Nonce
-            UserAgent.GetVarSize() +    //UserAgent
-            sizeof(uint);               //StartHeight
+            sizeof(uint) +              // Magic
+            sizeof(uint) +              // Version
+            sizeof(uint) +              // Timestamp
+            Capabilities.GetVarSize() + // Capabilities
+            sizeof(uint) +              // Nonce
+            UserAgent.GetVarSize();     // UserAgent
 
-        public static VersionPayload Create(int port, uint nonce, string userAgent, uint startHeight)
+        public static VersionPayload Create(uint nonce, string userAgent, IEnumerable<NodeCapabilityBase> capabilities)
         {
             return new VersionPayload
             {
                 Magic = ProtocolSettings.Default.Magic,
                 Version = LocalNode.ProtocolVersion,
-                Services = VersionServices.FullNode,
                 Timestamp = DateTime.Now.ToTimestamp(),
-                Port = (ushort)port,
                 Nonce = nonce,
                 UserAgent = userAgent,
-                StartHeight = startHeight,
+                Capabilities = capabilities.ToArray(),
             };
         }
 
         void ISerializable.Deserialize(BinaryReader reader)
         {
             Magic = reader.ReadUInt32();
             Version = reader.ReadUInt32();
-            Services = (VersionServices)reader.ReadUInt64();
             Timestamp = reader.ReadUInt32();
-            Port = reader.ReadUInt16();
+
+            // Capabilities
+
+            Capabilities = new NodeCapabilityBase[reader.ReadVarInt(MaxCapabilities)];
+
+            for (int x = 0, max = Capabilities.Length; x < max; x++)
+            {
+                Capabilities[x] = NodeCapabilityBase.Create((NodeCapabilities)reader.PeekChar());
```

Comment:
Not all streams support peeking.

Reply:
This could be a problem, because we need to check the type before...